src/ol/source/Vector.js

```diff
@@ -1116,22 +1116,36 @@ class VectorSource extends Source {
    */
   removeFeatureInternal(feature) {
     const featureKey = getUid(feature);
+    if (!this.uidIndex_.hasOwnProperty(featureKey)) {
+      return;
+    }
+
     if (featureKey in this.nullGeometryFeatures_) {
       delete this.nullGeometryFeatures_[featureKey];
     } else {
       if (this.featuresRtree_) {
         this.featuresRtree_.remove(feature);
       }
     }
+
     const featureChangeKeys = this.featureChangeKeys_[featureKey];
-    if (!featureChangeKeys) {
-      return;
+    if (featureChangeKeys) {
```

Comment:
We usually keep the happy path aligned as far left as possible, so it would be preferable to keep the logic that returns early when there ar no `featureChangeKeys`.

Reply:
But we can't return early because renderfeatures don't have change keys.
I could check for `!featureChangeKeys && feature instanceof Feature` though if you prefer

Reply:
Yes please, that would make the code easier to read.

Reply:
Actually, an early return also doesn't make sense here since `!featureChangeKeys && feature instanceof Feature` can never happen anymore due to the introduction of the `uidIndex` check at the top.
So I just removed the if clause and added optional chaining on the `featureChangeKeys.foreach`.